File: Programmers/연습문제/Level 1/신규 아이디 추천/Solution.py
import re


def solution(new_id):
    answer = ''
    
    step_0 = new_id
    step_1 = step_0.lower()
    step_2 = re.sub('[^a-z0-9-_.]', '', step_1)
    step_3 = re.sub('[.]+', '.', step_2)
    step_4 = re.sub('^[.]|[.]$', '', step_3)
    step_5 = 'a' if not step_4 else step_4
    step_6 = step_5[:15] if len(step_5) >= 16 else step_5
    step_6 = re.sub('[.]$', '', step_6)
    step_7 = step_6 if len(step_6) > 2 else step_6 + step_6[-1]*(3-len(step_6))
    
    answer = step_7
    return answer

File: Programmers/연습문제/Level 1/신규 아이디 추천/Solution.swift
import Foundation

func solution(_ new_id:String) -> String {
    let id1 = step1(new_id)
    let id2 = step2(id1)
    let id3and4 = step3and4(id2)
    let id5 = step5(id3and4)
    let id6 = step6(id5)
    let id7 = step7(id6)
    return id7
}

func step1(_ id: String) -> String {
    return id.lowercased()
}

func step2(_ id: String) -> String {
    let alphaRange = ("a"..."z")
    let symbolArray = ["-", "_", "."]
    return id.filter {  alphaRange.contains(String($0)) || $0.isNumber || symbolArray.contains(String($0)) }
}

func step3and4(_ id: String) -> String {
    return id.split(separator: ".").joined(separator: ".")
}

func step5(_ id: String) -> String {
    return id.isEmpty ? "a" : id
}

func step6(_ id: String) -> String {
    if id.count <= 15 {
        return id
    }
    
    let startIndex = id.startIndex
    let endIndex = id.index(id.startIndex, offsetBy: 15)
    var new_id = String(id[startIndex..<endIndex])
    if new_id.last! == "." {
        new_id = String(new_id.dropLast())
    }
    return new_id
}

func step7(_ id: String) -> String {
    if id.count > 2 {
        return id
    }
    
    let lastChar = id.last!
    return id + String(repeating: lastChar, count: 3 - id.count)
}



// TC
let new_id = "...!@BaT#*..y.abcdefghijklm"
print(solution(new_id)) // "bat.y.abcdefghi"
